File: 3-weeks-plan/week2-streaming-airflow/kafka/InMemoryKafkaBroker.java
import java.io.FileWriter;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.*;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

class InMemoryKafkaBroker {
    private final Map<String, BlockingQueue<String>> topics = new HashMap<>();
    private boolean active = false;

    public synchronized void start() {
        active = true;
        System.out.println("[InMemoryKafkaBroker] Started.");
    }
    public synchronized void stop() {
        active = false;
        System.out.println("[InMemoryKafkaBroker] Stopped.");
    }
    private synchronized void createTopic(String topic) {
        topics.computeIfAbsent(topic, k -> new LinkedBlockingQueue<>());
    }
    public void send(String topic, String value) {
        synchronized (this) {
            if (!active)
                throw new RuntimeException("Broker not started.");
            createTopic(topic);
            topics.get(topic).offer(value);
        }
    }
    public String consume(String topic, int timeoutMillis) {
        createTopic(topic);
        try {
            return topics.get(topic).poll(timeoutMillis, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            return null;
        }
    }
    public synchronized boolean isActive() { return active; }
}

class Producer implements Runnable {
    private final InMemoryKafkaBroker broker;
    private final String topic;
    private final int intervalSeconds;
    private static final String[] assets = {"USD/BDT", "EUR/USD", "BTC/USDT"};
    private final Random rand = new Random();
    private final Gson gson = new Gson();

    public Producer(InMemoryKafkaBroker broker, String topic, int intervalSeconds) {
        this.broker = broker; this.topic = topic; this.intervalSeconds = intervalSeconds;
    }

    public void run() {
        while (broker.isActive()) {
            Map<String, Object> trade = new HashMap<>();
            trade.put("market", "LocalMoneyExchange");
            trade.put("asset", assets[rand.nextInt(assets.length)]);
            trade.put("price", Math.round((rand.nextDouble() * 40 + 80) * 100.0) / 100.0);
            trade.put("volume", rand.nextInt(9901) + 100);
            trade.put("timestamp", System.currentTimeMillis() / 1000.0);
            String msg = gson.toJson(trade);
            broker.send(topic, msg);
            System.out.println("[Producer] Produced: " + msg);
            try { Thread.sleep(intervalSeconds * 1000); } catch (InterruptedException ignored) {}
        }
    }
}

class Consumer implements Runnable {
    private final InMemoryKafkaBroker broker;
    private final String topic;
    private final String dumpFile;
    private final List<Map<String, Object>> memoryData = new ArrayList<>();
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public Consumer(InMemoryKafkaBroker broker, String topic, String dumpFile) {
        this.broker = broker; this.topic = topic; this.dumpFile = dumpFile;
    }

    public void run() {
        while (broker.isActive()) {
            String msg = broker.consume(topic, 5000);
            if (msg != null) {
                Map<String, Object> trade = gson.fromJson(msg, Map.class);
                System.out.println("[Consumer] Consumed: " + msg);
                memoryData.add(trade);
                try (FileWriter file = new FileWriter(dumpFile)) {
                    gson.toJson(memoryData, file);
                } catch (IOException ignored) {}
            } else {
                try { Thread.sleep(1000); } catch (InterruptedException ignored) {}
            }
        }
    }
}

public class Main {
    public static void main(String[] args) throws Exception {
        InMemoryKafkaBroker broker = new InMemoryKafkaBroker();
        String topic = "money-market";
        broker.start();

        Thread producerThread = new Thread(new Producer(broker, topic, 10));
        Thread consumerThread = new Thread(new Consumer(broker, topic, "memory_bootstrap.json"));
        producerThread.setDaemon(true);
        consumerThread.setDaemon(true);

        producerThread.start();
        consumerThread.start();

        System.out.println("Press Enter to stop...");
        System.in.read();

        broker.stop();
        producerThread.join(2000);
        consumerThread.join(2000);
        System.out.println("[Main] Shutdown complete.");
    }
}

File: 3-weeks-plan/week2-streaming-airflow/kafka/InMemoryKafkaBroker.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

class InMemoryKafkaBroker
{
    private readonly Dictionary<string, ConcurrentQueue<string>> topics = new();
    private bool active = false;
    private readonly object lockObj = new();

    public void Start()
    {
        lock (lockObj)
        {
            active = true;
            Console.WriteLine("[InMemoryKafkaBroker] Started.");
        }
    }

    public void Stop()
    {
        lock (lockObj)
        {
            active = false;
            Console.WriteLine("[InMemoryKafkaBroker] Stopped.");
        }
    }

    public void CreateTopic(string topic)
    {
        lock (lockObj)
        {
            if (!topics.ContainsKey(topic))
                topics[topic] = new ConcurrentQueue<string>();
        }
    }

    public void Send(string topic, string value)
    {
        lock (lockObj)
        {
            if (!active)
                throw new InvalidOperationException("Broker not started.");
            CreateTopic(topic);
            topics[topic].Enqueue(value);
        }
    }

    public string Consume(string topic, int timeoutMillis = 1000)
    {
        DateTime end = DateTime.Now.AddMilliseconds(timeoutMillis);
        while (DateTime.Now < end)
        {
            lock (lockObj)
            {
                if (!active)
                    throw new InvalidOperationException("Broker not started.");
                CreateTopic(topic);
                if (topics[topic].TryDequeue(out var result))
                    return result;
            }
            Thread.Sleep(100);
        }
        return null;
    }

    public bool IsActive()
    {
        lock (lockObj) { return active; }
    }
}

class Program
{
    static string[] assets = { "USD/BDT", "EUR/USD", "BTC/USDT" };

    static void MoneyExchangeProducer(InMemoryKafkaBroker broker, string topic, int intervalSeconds)
    {
        var rand = new Random();
        while (broker.IsActive())
        {
            var trade = new
            {
                market = "LocalMoneyExchange",
                asset = assets[rand.Next(assets.Length)],
                price = Math.Round(rand.NextDouble() * 40 + 80, 2),
                volume = rand.Next(100, 10001),
                timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            };
            string msg = JsonSerializer.Serialize(trade);
            broker.Send(topic, msg);
            Console.WriteLine($"[Producer] Produced: {msg}");
            Thread.Sleep(intervalSeconds * 1000);
        }
    }

    static void MoneyExchangeConsumer(InMemoryKafkaBroker broker, string topic, string dumpFile)
    {
        var memoryData = new List<object>();
        while (broker.IsActive())
        {
            var msg = broker.Consume(topic, 5000);
            if (msg != null)
            {
                var trade = JsonSerializer.Deserialize<Dictionary<string, object>>(msg);
                Console.WriteLine($"[Consumer] Consumed: {msg}");
                memoryData.Add(trade);
                File.WriteAllText(dumpFile, JsonSerializer.Serialize(memoryData, new JsonSerializerOptions { WriteIndented = true }));
            }
            else
            {
                Thread.Sleep(1000);
            }
        }
    }

    static void Main()
    {
        var broker = new InMemoryKafkaBroker();
        string topic = "money-market";
        broker.Start();

        var producerThread = new Thread(() => MoneyExchangeProducer(broker, topic, 10));
        var consumerThread = new Thread(() => MoneyExchangeConsumer(broker, topic, "memory_bootstrap.json"));

        producerThread.Start();
        consumerThread.Start();

        Console.WriteLine("Press Enter to stop...");
        Console.ReadLine();

        broker.Stop();
        producerThread.Join(2000);
        consumerThread.Join(2000);
        Console.WriteLine("[Main] Shutdown complete.");
    }
}
